src/callbacks/callback_save_topk_models.cpp

```diff
@@ -0,0 +1,98 @@
+// https://github.com/LLNL/LBANN.
+//
+// Licensed under the Apache License, Version 2.0 (the "Licensee"); you
+// may not use this file except in compliance with the License.  You may
+// obtain a copy of the License at:
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+// implied. See the License for the specific language governing
+// permissions and limitations under the license.
+//
+// lbann_callback_save_topk_models .hpp .cpp - Callback hooks to save_topk_models information
+////////////////////////////////////////////////////////////////////////////////
+
+#include <vector>
+#include "lbann/callbacks/callback_save_topk_models.hpp"
+
+namespace lbann {
+void lbann_callback_save_topk_models::on_test_end(model *m) {
+  bool am_in_topk = false;
+  if(m->get_comm()->am_trainer_master()) {
+    am_in_topk = compute_stats(m);
+  }
+  m->get_comm()->trainer_broadcast(0, am_in_topk); 
+  if(am_in_topk) save_model(m);
+}
+
+bool lbann_callback_save_topk_models::compute_stats(model *m) {
+  lbann_comm *comm = m->get_comm();
+  const int num_trainers = comm->get_num_trainers();
+  std::string mode_string = "test";
+  bool found_metric = false;
+  EvalType score = 0;
+  for (const auto& met : m->get_metrics()) {
+    if (met->name() == m_metric_name) {
+      found_metric = true;
+      score = met->get_mean_value(m->get_execution_mode());
+      break;
+    }
+  }
+  //sanity check
+  if (!found_metric) {
+    std::stringstream err;
+    err << __FILE__ << " " << __LINE__ << " :: "
+        << "could not find metric \"" << m_metric_name << "\""
+        << "in model \"" << m->get_name() << "\"";
+    LBANN_ERROR(err.str());
+  }
+
+  if (m_k > num_trainers) {
+    std::stringstream err;
+    err << __FILE__ << " " << __LINE__ << " :: "
+        << "k ( " << m_k << ") " 
+        << " can not be greater than number of trainers (" 
+        << num_trainers << ") " ;
+    LBANN_ERROR(err.str());
+  }
+
+  std::vector<EvalType> score_list(comm->get_num_trainers());
+  comm->all_gather<EvalType>(score, score_list,comm->get_intertrainer_comm());
+  std::vector<EvalType> score_v = score_list;
+  //top-k in an ascending order
+  if(m_ascending_ordering) std::sort(score_v.begin(), score_v.end(),std::less<EvalType>());
+  //top-k in an descending order
+  else  std::sort(score_v.begin(), score_v.end(),std::greater<EvalType>());
+  score_v.resize(m_k);
+  
+  if (comm->am_world_master()) {
+    std::cout << "Top " << m_k << " " << m_metric_name << " average "
+              << std::accumulate(score_v.begin(), score_v.end(), EvalType(0))/m_k << std::endl;
+  } 
+  for(int i =0; i < num_trainers; ++i) {
+    if(std::find(score_v.begin(), score_v.end(), 
+                 score_list[i]) != score_v.end()) { 
```

Comment:
If multiple trainers have the same score, more than k trainers may be saved. In the extreme case, every trainer will be saved if all trainers have the same score. If we want to handle this edge case, we need to do the sorting more carefully.

If we don't mind the edge case and leave it as it is, we don't need to iterate through all the trainers. We only need to check if this trainer's score is in the top-k.

Reply:
The evaluation metric is just the first pass, so it is ok to save all the "good models" that meet initial condition for post-processing/ further evaluation (e.g., physics constraints)